\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1in]{geometry}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdftitle={An Overview of TidyGEO and Some Helpful Resources},
            pdfauthor={Avery Bell},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\providecommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

  \title{An Overview of TidyGEO and Some Helpful Resources}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
    \author{Avery Bell}
    \preauthor{\centering\large\emph}
  \postauthor{\par}
      \predate{\centering\large\emph}
  \postdate{\par}
    \date{June 25, 2019}


\begin{document}
\maketitle

\hypertarget{contents}{\subsection{Table of Contents}\label{contents}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  \textbf{\protect\hyperlink{intro-to-tidygeo}{Introduction to TidyGEO}}
\end{enumerate}

\begin{itemize}
\tightlist
\item
  \protect\hyperlink{intro-to-geo}{Introduction to Gene Expression
  Omnibus}
\item
  \protect\hyperlink{app-motivation}{Motivation for the app}
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  \textbf{\protect\hyperlink{use-case}{Use Case}}
\end{enumerate}

\begin{itemize}
\tightlist
\item
  \protect\hyperlink{geo-cleaning-steps}{Cleaning up a GEO dataset}
\item
  \protect\hyperlink{example}{Possible application}
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  \textbf{\protect\hyperlink{shiny-apps}{Shiny Apps}}
\end{enumerate}

\begin{itemize}
\tightlist
\item
  \protect\hyperlink{shiny-apps-overview}{Basic format}
\item
  \protect\hyperlink{shiny-apps-resources}{Resources}
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{3}
\tightlist
\item
  \textbf{\protect\hyperlink{tidygeo-code}{TidyGEO Code}}
\end{enumerate}

\begin{itemize}
\tightlist
\item
  \protect\hyperlink{directory-structure}{Directory structure}
\item
  \protect\hyperlink{file-contents}{File contents}
\item
  \protect\hyperlink{style-guide}{Style guide}
\item
  \protect\hyperlink{tidygeo-functions}{Functions and how to use them}
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{4}
\tightlist
\item
  \textbf{\protect\hyperlink{next-steps}{Next Steps}}
\end{enumerate}

\begin{itemize}
\tightlist
\item
  \protect\hyperlink{known-bugs}{Known bugs}
\item
  \protect\hyperlink{todo-items}{TODO items}
\item
  \protect\hyperlink{future-features}{Possible future features}
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{4}
\tightlist
\item
  \textbf{\protect\hyperlink{misc}{Miscellaneous}}
\end{enumerate}

\begin{itemize}
\tightlist
\item
  \protect\hyperlink{error-log}{Error log}
\item
  \protect\hyperlink{observations}{Shiny/R tips and observations}
\item
  \protect\hyperlink{time-testing}{Tips for time testing}
\item
  \protect\hyperlink{other-resources}{Other resources}

  \begin{itemize}
  \tightlist
  \item
    \protect\hyperlink{Rmarkdown-resource}{On RMarkdown}
  \end{itemize}
\end{itemize}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\hypertarget{intro-to-tidygeo}{\subsection{Introduction}\label{intro-to-tidygeo}}

\hypertarget{intro-to-geo}{\subsubsection{Gene Expression
Omnibus}\label{intro-to-geo}}

An online repository of array- and sequence-based data. Scientists can
go here for publicly available data to look at genes, proteins, etc.
that interest them. One of the largest databases of its kind.
Submissions are
\href{https://www.ncbi.nlm.nih.gov/geo/info/MIAME.html}{MIAME} compliant
but the metadata associated with the submissions (i.e., descriptions of
the sample sources, such as patient age and disease assessment) are not
regulated, so they are often messy and hard to analyse using
computational tools.

GEO data can be accessed via an R package called GEOQuery, but
scientists with no R coding experience must learn the language in order
to use this tool, and manual data extraction can often be tedious and
error-prone, especially with large batches of datasets that are
necessary for certain computational tools to perform their analysis.

There are some tools that have attempted to clean up the metadata and to
enable scientists to access it, but these tools either perform the
analysis for the user (not giving them access to the cleaned-up data) or
rely on crowd-curating efforts, which are not able to keep up with the
growing number of datasets on GEO.

\hypertarget{app-motivation}{\subsubsection{TidyGEO}\label{app-motivation}}

TidyGEO is an app that lets scientists (even those with little coding
experience) download and clean up GEO data to use for their own
purposes.

\protect\hyperlink{contents}{â†‘ To top}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\hypertarget{use-case}{\subsection{Use case}\label{use-case}}

\href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112914}{GSE112914}:
Differential gene expression analysis of MYCN-amplified neuroblastoma
cells after Doxycycline inducible shRNA knockdown of JMJD6 gene
expression

\subsubsection{\texorpdfstring{\href{https://brb.nci.nih.gov/BRB-ArrayTools/}{BRB
Array Tools}}{BRB Array Tools}}\label{brb-array-tools}

A package that helps visualize and perform statistical analysis on
Microarray gene expression, copy number, methylation and RNA-Seq data.

\hypertarget{geo-cleaning-steps}{\subsubsection{Cleaning up the
data}\label{geo-cleaning-steps}}

\href{https://tidygeo.shinyapps.io/tidygeo/}{TidyGEO}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Filter out unnecessary columns (to be less overwhelming)
\item
  Split key-value pairs
\item
  Split multiple values in a column
\item
  Shift cells
\item
  Substitute doxy and control in replicate for NA
\item
  Substitute ``repeat'' for ``''
\item
  Filter out NAs in replicate, or maybe duplicates in replicate
\item
  Download clinical data
\item
  Briefly look at assay and feature data
\end{enumerate}

\hypertarget{example}{\subsubsection{Analysis}\label{example}}

Now you can look at the effect of replicates on the expression
measurements.

\protect\hyperlink{contents}{â†‘ To top}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\hypertarget{shiny-apps}{\subsection{Shiny Apps}\label{shiny-apps}}

Shiny is an R package that makes it easy to build interactive web apps
straight from R.

\hypertarget{shiny-apps-overview}{\subsubsection{The
basics}\label{shiny-apps-overview}}

\begin{itemize}
\tightlist
\item
  \href{https://shiny.rstudio.com/tutorial/written-tutorial/lesson1/}{Structure
  of a Shiny App}
\item
  UI

  \begin{itemize}
  \tightlist
  \item
    The UI lays out what the app looks like. It is made up of
    \href{https://shiny.rstudio.com/gallery/widget-gallery.html}{widgets}
    that allow the user to interact with the app and
    \href{https://shiny.rstudio.com/articles/layout-guide.html}{layouts}
    that define how the widgets are arranged on the page.
  \end{itemize}
\item
  Server

  \begin{itemize}
  \tightlist
  \item
    The server is a function made up of

    \begin{itemize}
    \tightlist
    \item
      Outputs that define how UI objects should behave (e.g.,
      renderText, renderTable, etc.)
    \item
      Listeners that react to a user's input (e.g., observeEvent)
    \end{itemize}
  \item
    Basically, it does the \textbf{work} of the app to actually make it
    interactive
  \end{itemize}
\item
  Running an app

  \begin{itemize}
  \tightlist
  \item
    Call to shinyApp function
  \item
    Must have either \texttt{app.R} with UI and server function or
    \texttt{ui.R} and \texttt{server.R}
  \item
    Builds the app in the directory you name or the working directory,
    if no name is given
  \end{itemize}
\item
  Other resources

  \begin{itemize}
  \tightlist
  \item
    reactive values

    \begin{itemize}
    \tightlist
    \item
      the difference between reactive values and global variables
    \end{itemize}
  \item
    www/ directory

    \begin{itemize}
    \tightlist
    \item
      must be named www/, recognized by the app as a place where
      resources (graphics, CSS, JS scripts) are kept
    \end{itemize}
  \item
    CSS file

    \begin{itemize}
    \tightlist
    \item
      Cascading Style Sheets: an HTML file that defines how the app
      should look
    \item
      Shiny automatically uses Bootstrap styling but you can customize
      this in the CSS file
    \item
      generally it is good practice to use a CSS file rather than
      defining colors, sizes, etc. inline
    \end{itemize}
  \item
    JS scripts

    \begin{itemize}
    \tightlist
    \item
      for more complicated reactivity, Shiny may not be flexible enough
      to meet your needs
    \item
      you can write JavaScript that will define how certain objects
      react to button clicks, etc.
    \item
      this can also be defined inline but it is good practice to put
      this in a separate file
    \end{itemize}
  \item
    helper functions

    \begin{itemize}
    \tightlist
    \item
      code that is run in multiple parts of the app can be put in a
      function to reduce repetition
    \item
      often used by the server function to do work, but also may save
      some space in the UI part
    \item
      these are often kept in a separate file to clean up code
    \end{itemize}
  \item
    graphics

    \begin{itemize}
    \tightlist
    \item
      kept in the www/ directory
    \item
      may include browser icon, logos, icons, etc.
    \end{itemize}
  \end{itemize}
\end{itemize}

\hypertarget{shiny-apps-resources}{\subsubsection{Resources}\label{shiny-apps-resources}}

\begin{itemize}
\tightlist
\item
  \href{https://www.rstudio.com/resources/webinars/introduction-to-shiny/}{Introduction
  to Shiny video}
\item
  \href{https://shiny.rstudio.com/tutorial/}{Shiny Video and Written
  Tutorials}
\item
  \href{https://style.tidyverse.org/}{R Style Guide}
\end{itemize}

\protect\hyperlink{contents}{â†‘ To top}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\hypertarget{tidygeo-code}{\subsection{TidyGEO
Code}\label{tidygeo-code}}

\hypertarget{directory-structure}{\subsubsection{TidyGEO
structure}\label{directory-structure}}

\begin{itemize}
\tightlist
\item
  Root

  \begin{itemize}
  \tightlist
  \item
    \texttt{.gitignore}

    \begin{itemize}
    \tightlist
    \item
      Lists files that should not be updated on GitHub
    \end{itemize}
  \item
    \texttt{Code\_Review\_Notes.Rmd}

    \begin{itemize}
    \tightlist
    \item
      Some notes from my Code Review--pretty much the same as this
      document except less detailed
    \end{itemize}
  \item
    \texttt{geocurate\_repo.Rproj}

    \begin{itemize}
    \tightlist
    \item
      Some settings for the old Rproject (an RStudio thing)
    \end{itemize}
  \item
    \texttt{README.md}

    \begin{itemize}
    \tightlist
    \item
      The readme for the GitHub repo
    \end{itemize}
  \item
    \texttt{TidyGEO/}

    \begin{itemize}
    \tightlist
    \item
      Contains all the code
    \end{itemize}
  \item
    \texttt{TidyGEO\_Directory\_Guide.Rmd}
  \item
    This file
  \end{itemize}
\item
  Level 1 - TidyGEO

  \begin{itemize}
  \tightlist
  \item
    \texttt{app.R}

    \begin{itemize}
    \tightlist
    \item
      sources all the R scripts from the ui and server directories
    \item
      creates the ui object and the server function
    \item
      creates some lists of reactiveValues that are used in the app
    \item
      contains some instructions for how column navigation should work
    \item
      contains shinyApp function to knit UI and server
    \end{itemize}
  \item
    \texttt{deploy\_app.R}

    \begin{itemize}
    \tightlist
    \item
      updates the list of available GEO series to choose from
    \item
      updates the list of formatting functions that can be included in
      the scripts the user can download
    \item
      launches the app to shinyapps.io
    \end{itemize}
  \item
    \texttt{error\_log.md}

    \begin{itemize}
    \tightlist
    \item
      a list of errors I have encountered and how I fixed them
    \end{itemize}
  \item
    \texttt{file\_manifest.txt}

    \begin{itemize}
    \tightlist
    \item
      defines which files are necessary when publishing the app to the
      web
    \end{itemize}
  \item
    \texttt{generate\_profile.R}

    \begin{itemize}
    \tightlist
    \item
      generates a profile for the app, see the
      \protect\hyperlink{time-testing}{time testing tips} section
    \end{itemize}
  \item
    \texttt{generate\_rscript\_functions.R}

    \begin{itemize}
    \tightlist
    \item
      generates a list of functions that can be used in the scripts the
      user can download
    \end{itemize}
  \item
    \texttt{tidygeo-install.R}

    \begin{itemize}
    \tightlist
    \item
      contains the packages needed to run the app and use any of the
      functions in the app (some packages may be extraneous--this hasn't
      been updated in a while)
    \end{itemize}
  \item
    \texttt{tidygeo\_functions.R}

    \begin{itemize}
    \tightlist
    \item
      some functions for the app that don't really fall under the
      ``ui'', ``server'', or ``variables'' categories; mostly for script
      writing
    \end{itemize}
  \item
    \texttt{tidygeo\_variables.R}

    \begin{itemize}
    \tightlist
    \item
      consists of two parts: global variables and naming conventions
    \item
      global variables are constants used by multiple parts of the app
    \item
      naming conventions are patterns for which variables in the app are
      named; these are useful because the same name will often be
      referenced by both the server and the ui portions, so naming
      conventions make it so you don't have to remember as many variable
      names; these are also useful because similar variables will often
      appear for each of the datatypes (clinical, assay, feature, all)
      and naming conventions make it so you can name these all similarly
      and reference each of them without having to remember how you
      named them
    \end{itemize}
  \item
    \texttt{time\_testing.R}

    \begin{itemize}
    \tightlist
    \item
      basically a sandbox for optimizing various pieces of code; se the
      \protect\hyperlink{time-testing}{time testing tips} section
    \end{itemize}
  \item
    \texttt{VERSION}

    \begin{itemize}
    \tightlist
    \item
      defines the version of the app, updates with every commit
    \end{itemize}
  \item
    \texttt{help\_docs/}

    \begin{itemize}
    \tightlist
    \item
      Markdown files for the help icons in the app
    \item
      (Generally it's a good idea to separate out large portions of text
      from code that actually does work.)
    \end{itemize}
  \item
    \texttt{rsconnect/}

    \begin{itemize}
    \tightlist
    \item
      an automatically-generated directory that I do not touch;
      necessary to publish the app to shinyapps.io
    \end{itemize}
  \item
    \texttt{series\_platform/}

    \begin{itemize}
    \tightlist
    \item
      contains a script that populates the dropdown you see on the first
      tab
    \end{itemize}
  \item
    \texttt{server/}

    \begin{itemize}
    \tightlist
    \item
      contains all the code for the server function
    \item
      subsectioned by datatype (clinical, assay, feature, all)
    \end{itemize}
  \item
    \texttt{ui/}

    \begin{itemize}
    \tightlist
    \item
      contains all the code for the ui object
    \item
      subsectioned by datatype (clinical, assay, feature, all)
    \end{itemize}
  \item
    \texttt{User/}

    \begin{itemize}
    \tightlist
    \item
      pretty much everything in this directory is useless except
      rscript\_functions.rds--it is the list of the functions that can
      be included in the script the user can download
    \end{itemize}
  \item
    \texttt{www/}

    \begin{itemize}
    \tightlist
    \item
      resources for the app, including help gifs, the series and
      platform lists, the javascript file, and the css styling file
    \end{itemize}
  \end{itemize}
\item
  Level 2 - \texttt{server}

  \begin{itemize}
  \tightlist
  \item
    Each directory contains R scripts organized by tab (the name of the
    tab is usually the name of the R script, and the tab is named by
    what the user does in that tab)
  \item
    \texttt{assay/}

    \begin{itemize}
    \tightlist
    \item
      \texttt{feature\_data.R} is the code for the modal that comes up
      when you click the ``Replace ID'' button
    \item
      \texttt{side\_panel.R} and \texttt{helper\_functions.R} are pretty
      much useless (the code in helper\_functions.R should all be in
      \texttt{formatting\_functions.R} now)
    \end{itemize}
  \item
    \texttt{clinical/}

    \begin{itemize}
    \tightlist
    \item
      \texttt{helper\_functions.R} is pretty much useless (the code in
      \texttt{helper\_functions.R} should all be in
      \texttt{formatting\_functions.R} now)
    \end{itemize}
  \item
    \texttt{formatting\_helper\_functions.R}

    \begin{itemize}
    \tightlist
    \item
      the functions that can be included in the script the user can
      download; \textbf{these same functions} are used in the app to
      format the data in various ways--usually one function per tab
    \end{itemize}
  \item
    \texttt{help\_modals.R}

    \begin{itemize}
    \tightlist
    \item
      functions to create the modals that come up when the user clicks a
      question icon, observers to make the question icons reactive
    \end{itemize}
  \item
    \texttt{regex\_modal.R}

    \begin{itemize}
    \tightlist
    \item
      the code to create the modal that allows users to test regular
      expressions
    \end{itemize}
  \item
    \texttt{tidygeo\_server\_functions.R}

    \begin{itemize}
    \tightlist
    \item
      functions to create various pieces of the server; these often
      correspond to similarly-named ui functions to create an object the
      user can interact with
    \end{itemize}
  \end{itemize}
\item
  Level 2 - \texttt{ui}

  \begin{itemize}
  \tightlist
  \item
    Each directory contains R scripts organized by tab (the name of the
    tab is usually the name of the R script, and the tab is named by
    what the user does in that tab)
  \item
    \texttt{assay/}

    \begin{itemize}
    \tightlist
    \item
      \texttt{side\_panel.R} and \texttt{side\_panel\_assay.R} are
      useless
    \end{itemize}
  \item
    \texttt{feature/}

    \begin{itemize}
    \tightlist
    \item
      \texttt{side\_panel\_feature.R} is useless
    \end{itemize}
  \item
    \texttt{tidygeo\_ui\_functions.R}

    \begin{itemize}
    \tightlist
    \item
      functions to create various pieces of the ui; these often
      correspond to similarly-named server functions to create an object
      the user can interact with
    \end{itemize}
  \end{itemize}
\item
  Level 2 - \texttt{www}

  \begin{itemize}
  \tightlist
  \item
    \texttt{reactive\_preferences.js}

    \begin{itemize}
    \tightlist
    \item
      when I have a bunch of buttons that do the same thing, I can't
      name the buttons the same thing (duplicate IDs not allowed) and I
      can't have an \texttt{observeEvent} that listens to multiple
      buttons; instead, I give the buttons a class and whenever a button
      from that class is clicked, the JavaScript here puts that button's
      ID in an input for me to listen to with an \texttt{observeEvent};
      after I've used that input, I have to reset it with this
      JavaScript so the next button click can set off the
      \texttt{observeEvent} again
    \item
      it's generally good practice to put any JS in a separate script
      such as this one so that your R code doesn't look as cluttered
    \end{itemize}
  \item
    \texttt{style.css}

    \begin{itemize}
    \tightlist
    \item
      a bunch of HTML that defines how certain elements of the app
      should look
    \item
      for each chunk, you'll see a dot followed by the name of a class
      followed by some attributes for the class
    \item
      it's generally good practice to put any styling HTML in a separate
      script such as this one so that your R code doesn't look as
      cluttered
    \end{itemize}
  \end{itemize}
\end{itemize}

\hypertarget{style-guide}{\subsubsection{Style guide and naming
practices}\label{style-guide}}

I tried my best to follow the tidyverse style guide written by Hadley
Wickham as it appears in \href{https://style.tidyverse.org/}{this
article}.

\paragraph{Naming conventions}\label{naming-conventions}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Most variable and function names are all undercase with underscores
  separating the words. If you ever want to change this convention,
  please change the \texttt{SEP} variable in
  \texttt{tidygeo\_variables.R}.
\item
  Much of the app is built using functions, and these functions use
  specific names to reference certain variable names in the app. Rather
  than having the programmer memorize how the functions reference the
  variable names, I created naming patterns in
  \texttt{tidygeo\_variables.R} that define how certain types of
  variables are named. To change the naming pattern, one simply changes
  \texttt{tidygeo\_variables.R} and all the variable names in the app
  will be updated. In this way, the naming conventions in
  \texttt{tidygeo\_variables.R} also prevent extensive refactoring.
\item
  In general, variables are named using three parts: the general name
  for the variable, the datatype the variable corresponds to (clinical,
  assay, feature, all), and some kind of extra tag to make that variable
  unique, in case it is used for the same datatype in multiple sections.
\end{enumerate}

\paragraph{Organization}\label{organization}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Tabs are organized by datatype and by what the user can do to that
  data in the tab.
\item
  Code is usually organized by datatype and tab. Code chunks are
  separated by a comment ending in three or more dashes, which can be
  easily produced in RStudio using \texttt{Ctrl} + \texttt{Shift} +
  \texttt{R}.
\item
  Chunks of text for the app are usually kept in separate files, not in
  the R code.
\item
  All global variables are defined in \texttt{tidygeo\_variables.R}, not
  in app.R or any of the scripts that build the app.
\item
  All libraries needed for the app to run are loaded in
  \texttt{tidygeo\_variables.R}.
\item
  Functions are always defined outside the tab R scripts. Functions that
  are used in the UI portion are defined in
  \texttt{tidygeo\_ui\_functions.R}. Functions that make changes to the
  data are defined in \texttt{formatting\_functions.R}. Functions that
  are used in the Server portion are defined in
  \texttt{tidygeo\_server\_functions.R}. Functions that do not fit into
  a category are defined in \texttt{tidygeo\_functions.R}.
\end{enumerate}

\hypertarget{tidygeo-functions}{\subsubsection{Functions and how to use
them}\label{tidygeo-functions}}

There are two reasons that I would \textbf{strongly encourage} any
programmer working on this app to use the existing functions rather than
writing things from scratch.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Pieces of UI often correspond to similarly-named pieces of Server.
\end{enumerate}

Just like you pair a \texttt{textOutput} in the UI with a
\texttt{renderText} in the Server, I have created pieces of UI that
correspond to pieces of the Server, and these corresponding pieces have
similar names. One example is with the column navigation you might
notice above each of the datatables in the app. The UI for the column
navigation looks like this:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{fluidRow}\NormalTok{( }\CommentTok{# A row for all these elements}
    \KeywordTok{column}\NormalTok{(}\DecValTok{4}\NormalTok{, }\CommentTok{# The "previous columns" button}
      \KeywordTok{secondary_button}\NormalTok{(}\DataTypeTok{id =} \StringTok{"prev_cols_clinical_data_viewer"}\NormalTok{, }\DataTypeTok{label =} \KeywordTok{div}\NormalTok{(PREV_ICON, }\StringTok{"Previous columns"}\NormalTok{), }\DataTypeTok{class =} \StringTok{"prev_cols"}\NormalTok{)}
\NormalTok{    ),}
    \KeywordTok{column}\NormalTok{(}\DecValTok{4}\NormalTok{, }\CommentTok{# A spot for the text showing what subset the user is viewing}
      \KeywordTok{div}\NormalTok{(}\KeywordTok{textOutput}\NormalTok{(}\StringTok{"cols_visible_clinical_data_viewer"}\NormalTok{), }\DataTypeTok{class =} \StringTok{"center_align"}\NormalTok{)}
\NormalTok{    ),}
    \KeywordTok{column}\NormalTok{(}\DecValTok{4}\NormalTok{, }\CommentTok{# The "next columns" button}
      \KeywordTok{div}\NormalTok{(}\KeywordTok{secondary_button}\NormalTok{(}\DataTypeTok{id =} \StringTok{"next_cols_clinical_data_viewer"}\NormalTok{, }\DataTypeTok{label =} \StringTok{"Next columns"}\NormalTok{, }\DataTypeTok{icon =}\NormalTok{ NEXT_ICON, }\DataTypeTok{class =} \StringTok{"next_cols"}\NormalTok{), }\DataTypeTok{class =} \StringTok{"right_align"}\NormalTok{)}
\NormalTok{    )}
\NormalTok{  )}
\end{Highlighting}
\end{Shaded}

and the server looks like this:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# This fills in the spot with the text showing what subset the user is viewing}
\NormalTok{output[[}\StringTok{"cols_visible_clinical_data_viewer"}\NormalTok{]] <-}\StringTok{ }\KeywordTok{renderText}\NormalTok{(\{}
  \KeywordTok{paste}\NormalTok{(}\StringTok{"Showing"}\NormalTok{, }
\NormalTok{        clinical_vals}\OperatorTok{$}\NormalTok{viewing_subset[}\DecValTok{1}\NormalTok{], }\StringTok{"to"}\NormalTok{,}
\NormalTok{        clinical_vals}\OperatorTok{$}\NormalTok{viewing_subset[}\DecValTok{2}\NormalTok{], }\StringTok{"of"}\NormalTok{,}
        \KeywordTok{ncol}\NormalTok{(clinical_vals}\OperatorTok{$}\NormalTok{clinical_data),}
        \StringTok{"columns"}\NormalTok{)}
\NormalTok{\})}
\end{Highlighting}
\end{Shaded}

See how the variable names in the UI and Server look similar? They all
have the name of the element (``prev\_cols'', ``cols\_visble'', or
``next\_cols''), then the datatype, then the section where the element
appears (``data\_viewer''), all separated by an underscore. Now, rather
than repeating these chunks of UI and Server for every single datatype,
I made functions that will create these pieces for me:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# ui functions}
\CommentTok{# The "extra tag" specifies the section for the element; it helps make the element unique}
\NormalTok{col_navigation_set <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(datatype, }\DataTypeTok{extra_tag =} \OtherTok{NULL}\NormalTok{) \{}
  \KeywordTok{fluidRow}\NormalTok{(}
    \KeywordTok{column}\NormalTok{(}\DecValTok{4}\NormalTok{,}
      \KeywordTok{secondary_button}\NormalTok{(}\DataTypeTok{id =} \KeywordTok{prev_col}\NormalTok{(datatype, extra_tag), }\DataTypeTok{label =} \KeywordTok{div}\NormalTok{(PREV_ICON, }\StringTok{"Previous columns"}\NormalTok{), }\DataTypeTok{class =} \StringTok{"prev_cols"}\NormalTok{)}
\NormalTok{    ),}
    \KeywordTok{column}\NormalTok{(}\DecValTok{4}\NormalTok{,}
      \KeywordTok{div}\NormalTok{(}\KeywordTok{textOutput}\NormalTok{(}\KeywordTok{visible}\NormalTok{(datatype, extra_tag)), }\DataTypeTok{class =} \StringTok{"center_align"}\NormalTok{)}
\NormalTok{    ),}
    \KeywordTok{column}\NormalTok{(}\DecValTok{4}\NormalTok{,}
      \KeywordTok{div}\NormalTok{(}\KeywordTok{secondary_button}\NormalTok{(}\DataTypeTok{id =} \KeywordTok{next_col}\NormalTok{(datatype, extra_tag), }\DataTypeTok{label =} \StringTok{"Next columns"}\NormalTok{, }\DataTypeTok{icon =}\NormalTok{ NEXT_ICON, }\DataTypeTok{class =} \StringTok{"next_cols"}\NormalTok{), }\DataTypeTok{class =} \StringTok{"right_align"}\NormalTok{)}
\NormalTok{    )}
\NormalTok{  )}
\NormalTok{\}}
\CommentTok{# server functions}
\CommentTok{# Uses the same "extra tag" as above}
\NormalTok{col_navigation_set_server <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(datatype, }\DataTypeTok{extra_tag =} \OtherTok{NULL}\NormalTok{) \{}
\NormalTok{  output[[}\KeywordTok{visible}\NormalTok{(datatype, extra_tag)]] <-}\StringTok{ }\KeywordTok{renderText}\NormalTok{(\{}
    \KeywordTok{paste}\NormalTok{(}\StringTok{"Showing"}\NormalTok{, }
          \KeywordTok{get_data_member}\NormalTok{(datatype, }\StringTok{"viewing_subset"}\NormalTok{)[}\DecValTok{1}\NormalTok{], }\StringTok{"to"}\NormalTok{, }
          \KeywordTok{get_data_member}\NormalTok{(datatype, }\StringTok{"viewing_subset"}\NormalTok{)[}\DecValTok{2}\NormalTok{], }\StringTok{"of"}\NormalTok{, }
          \KeywordTok{ncol}\NormalTok{(}\KeywordTok{get_data_member}\NormalTok{(datatype, }\KeywordTok{dataname}\NormalTok{(datatype))), }
          \StringTok{"columns"}\NormalTok{)}
\NormalTok{  \})}
\NormalTok{\}}
\CommentTok{# ui}
\KeywordTok{col_navigation_set}\NormalTok{(}\StringTok{"clinical"}\NormalTok{, }\StringTok{"data_viewer"}\NormalTok{)}
\CommentTok{# server}
\KeywordTok{col_navigation_set_server}\NormalTok{(}\StringTok{"clinical"}\NormalTok{, }\StringTok{"data_viewer"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  The use of functions makes variable names very important
\end{enumerate}

You might have noticed how the functions above had to know a little bit
about how the variables would be named in order to reference them. For
example, the server function had to know that the textOutput spot would
start with ``cols\_visible'' and end in the datatype and an extra tag.
Rather than forcing the programmer to memorize this naming schema, I
created functions that replicate the naming schema--for this example, a
function called \texttt{visible} that creates the string
\texttt{cols\_visible\_clinical\_data\_viewer} given the datatype
\texttt{clinical} and the section \texttt{data\_viewer}. Functions like
this are also helpful when the programmer wants to change the naming
schema. For example, if the programmer wanted to change the prefix from
\texttt{cols\_visible} to \texttt{view\_subset}, they would simply need
to change one line in\texttt{tidygeo\_variables.R} rather than
refactoring all of the ui and server code.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  Functions have error-checking in place
\end{enumerate}

This app currently only has four datatypes: clinical, assay, feature,
and all, and if one references a datatype that does not exist, say they
tried to call \texttt{expression\_vals\$viewing\_subset}, they will get
an error. Referencing the datatypes by hand (i.e.
{[}datatype{]}\texttt{\_vals\$}{[}datatype{]}\texttt{\_data}) will throw
a generic ``variable not found'' error that may be difficult to track
down, but using functions (i.e.~get\_data\_member(datatype,
dataname(datatype))) will make use of my built-in error checking that
indicates which function caused the error and keeps the app from
crashing.

\protect\hyperlink{contents}{â†‘ To top}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\hypertarget{next-steps}{\subsection{Next Steps}\label{next-steps}}

\hypertarget{known-bugs}{\subsubsection{Known bugs}\label{known-bugs}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Loading assay data into the app is ridiculously slow because the
  loading function must perform the O(n\^{}2) operation of checking
  whether each cell is a blank string " " and replacing it with NA if
  so.
\item
  If the user wants to load the data into BRB array tools, the assay
  data can't have been transposed. There is currently nothing to prevent
  the user from doing this.
\item
  There's not really a great way to deal with invalid column names when
  the user renames a column to something like ``\#\$\%\^{}\&*()``.
  Currently the app just keeps the names, until any formatting is
  performed on the column, then the name might change to something
  like''V1\ldots{}.."
\item
  The wording for that ``drop NA'' checkbox in the feature modal in the
  assay data tab is kind of vague and confusing.
\item
  When the assay data is not numeric, the error message is unclear on
  which functionality is not supported by the app.
\end{enumerate}

\hypertarget{todo-items}{\subsubsection{TODO items}\label{todo-items}}

\hypertarget{future-features}{\subsubsection{Possible future
features}\label{future-features}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Sometimes, as is the case for GSE20181, datasets will have subsets of
  rows that should be treated differently from the other rows. Currently
  the app only supports edits on entire columns. Perhaps we could add a
  feature to select which rows the user would like to reformat.
\item
  There are no data analysis options for the assay data. We could look
  into some common data analysis things like the ones BRB Array Tools
  expects the data to already have when it is loaded.
\item
  There are very few formatting options for the feature data.
\item
  We could add options to create graphs besides bar plots/histograms.
\item
  We could add standardization options such as those that appear in
  GoodNomen.
\item
  We could suggest certain edits based on what we find in the data. For
  example, if a column has a non-word character, we could suggest that
  the user split columns on a delimiter.
\item
  If not all rows in a column contain a delimiter when the user wants to
  split the column, the entire column is not split. We could make this
  smarter so it only splits the rows that contain the delimiter and
  leaves the rest in the first column.
\item
  There's probably a better way to summarize the data so the user can
  see what's in all the rows than the bar plots. The labels in the bar
  plots are often cut off so it's hard to see the unique values.
\item
  We could add an option to transpose/summarize clinical data like in
  the assay data tab.
\item
  We could make it so the series and platform lists don't have to be
  udpated manually.
\item
  The rownames on the clinical data aren't really necessary since
  there's always a \texttt{geo\_accession} column. We could drop the
  rownames and make the \texttt{geo\_accession} column an ID like the
  IDs in the assay and feature data tabs.
\item
  We could add a way to combine two different series.
\end{enumerate}

\protect\hyperlink{contents}{â†‘ To top}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\hypertarget{misc}{\subsection{Miscellaneous}\label{misc}}

\hypertarget{error-log}{\subsubsection{Error log}\label{error-log}}

When I run into a tricky error, I make note of it in
\texttt{error\_log.md}, so when I encounter the error again, I can know
how to fix it quickly. Here are some of the errors I have encountered so
far.

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\subsection{Errors}\label{errors}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  When you try to click the download button after something has been
  downloaded already (after you click ``Undo''\ldots{} you can check if
  this is always the case), you get the following error:

\begin{verbatim}
Warning: Error in normalizeChoicesArgs: Please specify a non-empty vector for `choices` (or, alternatively, for both `choiceNames` AND `choiceValues`).
observeEventHandler [C:\Users\Avery\Documents\R_Code\geocurate_repo\GEOcurate/app.R#371]
\end{verbatim}

  \textbf{SOLVED} Kept the platforms module from showing if the
  platforms do not exist. ``Undo'' should work normally now.
\item
  After downloading GSE68849, if you click on the summary tab, it will
  throw the following error (perhaps it has something to do with the
  semicolons in the column names?):

\begin{verbatim}
Warning: Error in <Anonymous>: No handler registered for type .clientdata_output_agent:ch1_width
  [No stack trace available]
Error in (function (name, val, shinysession)  : 
  No handler registered for type .clientdata_output_agent:ch1_width
\end{verbatim}

  \textbf{SOLVED} Added make.names to the plotting code (Summary tab) to
  get rid of the illegal characters in the column names, but just for
  the back-end of the plots.
\item
  GSE10 Expression graphical summary:
  \texttt{Error:\ missing\ value\ where\ TRUE/FALSE\ needed}

  \textbf{SOLVED} Added \texttt{na.rm\ =\ TRUE} in the renderUI where
  it's figuring out the default width of the bars
\item
  \texttt{cannot\ remove\ prior\ installation\ of\ package\ \textquotesingle{}mime\textquotesingle{}}

  \textbf{SOLVED}

  \begin{itemize}
  \tightlist
  \item
    Screenshot the place where the packages are being installed
  \item
    End RStudio using Task Manager
  \item
    Navigate to the place where the packages are being installed
  \item
    Try to delete \texttt{mime} directory
  \item
    If the directory says it's being used by another program,
  \item
    Navigate to the innermost level of the directory
  \item
    Attempt to delete the innermost file
  \item
    The message will tell you which program is using the directory
  \item
    Find that program in Task Manager
  \item
    End the program
  \item
    Open RStudio again
  \item
    Reinstall the package
  \end{itemize}
\item
  \texttt{Error\ in\ value{[}{[}3L{]}{]}(cond)\ :\ there\ is\ no\ package\ called\ â€˜markdownâ€™}

  \textbf{SOLVED} Added \texttt{library(rmarkdown)} to app.R
  (\url{https://community.rstudio.com/t/error-in-deploying/33173/14})
\end{enumerate}

\subsection{Fixes}\label{fixes}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  The app does not allow the user to choose which new column name to
  remove.

  \textbf{LEFT} The ``remove'' button just takes off the last name that
  was specified, so if the user wants to delete a different one, they
  must start over.
\item
  The app allows the user to put multiple of the same column name.

  \textbf{SOLVED} Added unique = TRUE to make.names in the renameCols
  function.
\item
  When substituting values, the options of which values to substitute do
  not update after the substitute operation has been evaluated.

  \textbf{SOLVED} Updated the options after the ``Substitute'' button is
  pressed.
\item
  The feature data table does not have select inputs for any of the
  filters, even though those columns should be factors.

  \textbf{LEFT}

  \begin{itemize}
  \tightlist
  \item
    \url{https://stackoverflow.com/questions/49699991/how-to-change-the-column-filter-control-in-r-shiny-datatables}
  \item
    \url{https://rstudio.github.io/DT/}
  \end{itemize}
\item
  The app is taking too long to load because of the series list file.

  \textbf{SOLVED} Used feather files instead of RDS files to store the
  data.
  \texttt{{[}1{]}\ "RDS\ time:"\ \ \ \ \ \ \ \ \ "0.201472043991089"\ {[}1{]}\ "Feather\ time:"\ \ \ \ \ \ "0.0897789001464844"}
\item
  The app is still taking too long to load because of the series list
  file.

  \textbf{SOLVED} Prepared the feather files all beforehand (naming the
  columns correctly, including the correct columns) in another script,
  because making a dataframe in the list dropdown is very
  time-expensive.

\begin{verbatim}
[1] "Reading files 0.107712984085083"
[1] "Populating dropdown 0.00898313522338867"
\end{verbatim}
\end{enumerate}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\hypertarget{observations}{\subsubsection{Shiny/R tips and
observations}\label{observations}}

\paragraph{On renderUI vs updateInput}\label{on-renderui-vs-updateinput}

\texttt{renderUI}s are, in general, better than observes with
\texttt{update{[}{]}Input}s in them. This is because observes are called
more often--whenever their dependencies are updated--rather than only
when we need to use them in the UI. I have a hypothesis, though, that
with computationally-intensive UIs that we need to use often and that
update often, an \texttt{update{[}{]}Input} might outperform a
\texttt{renderUI} since it only has to update a part of the UI rather
than render the entire thing again. However, I have never encountered a
UI so difficult to render that \texttt{renderUI} lagged down the code.
So IDK.

\hypertarget{time-testing}{\subsubsection{Tips for time
testing}\label{time-testing}}

There is an R package called profvis that can perform profiling for
Shiny apps, i.e., it watches the app as the app runs and when the app
stops, outputs data on which parts of the code took the longest and
which were called the most. This package usually didn't work for me, so
I copy-and-pasted sections of code into \texttt{time\_testing.R}, used
Sys.time() calls before and after each section to check how long it
took, and printed out the time to see how many times it was called. All
the Sys.time() calls should be removed but there might be some I forgot
about.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{start <-}\StringTok{ }\KeywordTok{Sys.time}\NormalTok{()}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{10}\NormalTok{) \{}
  \KeywordTok{Sys.sleep}\NormalTok{(}\FloatTok{0.1}\NormalTok{)}
\NormalTok{\}}
\NormalTok{end <-}\StringTok{ }\KeywordTok{Sys.time}\NormalTok{()}
\KeywordTok{print}\NormalTok{(end }\OperatorTok{-}\StringTok{ }\NormalTok{start)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Time difference of 1.111573 secs
\end{verbatim}

\hypertarget{other-resources}{\subsubsection{Other
resources}\label{other-resources}}

\hypertarget{Rmarkdown-resource}{\paragraph{On
Rmarkdown}\label{Rmarkdown-resource}}

Pimp my RMD: a few tips for R Markdown, by Yan Holtz
\href{https://holtzy.github.io/Pimp-my-rmd/}{(link)}

\protect\hyperlink{contents}{â†‘ To top}


\end{document}
